### PR TITLE
[openebs/litmus] (test): Include a litmus book to perform node maintenance (via kubectl drain).

### PR DESCRIPTION
**Pull request by dargasudarshan**
Signed-off-by: Sudarshan <sudarshan.darga@cloudbyte.com>

**What this PR does / why we need it**:
- Include a litmus book to perform node maintenance
- This test requires a cluster with 4 nodes and one node tainted to run only liveness and litmus pods.
- Pre-chaos and post chaos will verify the application pod status by verifying the liveness pod status

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by dargasudarshan** [comment]
```
[0;34mansible-playbook 2.7.2[0m
[0;34m  config file = None[0m
[0;34m  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules'][0m
[0;34m  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible[0m
[0;34m  executable location = /usr/local/bin/ansible-playbook[0m
[0;34m  python version = 2.7.15rc1 (default, Nov 12 2018, 14:31:15) [GCC 7.3.0][0m
[0;34mNo config file found; using defaults[0m
[0;34m/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected[0m
[0;34m/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected[0m

PLAYBOOK: test.yml *************************************************************
[0;34m1 plays in ./percona/chaos/drain_node/test.yml[0m

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:2[0m
[0;32mok: [127.0.0.1][0m
[0;34mMETA: ran handlers[0m

TASK [Get the liveness pods] ***************************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:14[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod -n litmus -l liveness=node-chaos -o=custom-columns=NAME:\".metadata.name\" --no-headers", "delta": "0:00:01.733082", "end": "2018-11-26 11:56:07.397552", "rc": 0, "start": "2018-11-26 11:56:05.664470", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}[0m

TASK [Checking status of liveness pods] ****************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:18[0m

TASK [Record test instance/run ID] *********************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:32[0m
[0;36mskipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}[0m

TASK [Construct testname appended with runID] **********************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:36[0m
[0;36mskipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}[0m

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
[1;30mtask path: /percona/chaos/drain_node/test.yml:42[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "checksum": "0bc7dfd5b0edc01bdbca191f161bf7ef3ffaa7aa", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "6e1afe677980b0a0b1189a2f80338334", "mode": "0644", "owner": "root", "size": 436, "src": "/root/.ansible/tmp/ansible-tmp-1543233368.22-15935805992181/source", "state": "file", "uid": 0}[0m

TASK [Apply the litmus result CR] **********************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:53[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.991875", "end": "2018-11-26 11:56:13.262284", "failed_when_result": false, "rc": 0, "start": "2018-11-26 11:56:11.270409", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/node-failure-drain configured", "stdout_lines": ["litmusresult.litmus.io/node-failure-drain configured"]}[0m

TASK [Display the app information passed via the test job] *********************
[1;30mtask path: /percona/chaos/drain_node/test.yml:62[0m
[0;32mok: [127.0.0.1] => {[0m
[0;32m    "msg": [[0m
[0;32m        "The application info is as follows:", [0m
[0;32m        "Namespace    : app-percona-ns", [0m
[0;32m        "Label        : name=percona"[0m
[0;32m    ][0m
[0;32m}[0m

TASK [Verify that the AUT (Application Under Test) is running] *****************
[1;30mtask path: /percona/chaos/drain_node/test.yml:69[0m
[0;33mchanged: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n app-percona-ns -l name=percona --no-headers -o custom-columns=:status.phase", "delta": "0:00:01.905705", "end": "2018-11-26 11:56:16.802676", "rc": 0, "start": "2018-11-26 11:56:14.896971", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}[0m

TASK [Get application pod name] ************************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:80[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -n app-percona-ns -l name=percona --no-headers -o=custom-columns=NAME:\".metadata.name\"", "delta": "0:00:01.230835", "end": "2018-11-26 11:56:19.016510", "rc": 0, "start": "2018-11-26 11:56:17.785675", "stderr": "", "stderr_lines": [], "stdout": "percona-6966c6db86-8lwfb", "stdout_lines": ["percona-6966c6db86-8lwfb"]}[0m

TASK [include_tasks] ***********************************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:90[0m
[0;36mincluded: /chaoslib/kubectl/cordon_drain_node.yaml for 127.0.0.1[0m

TASK [Identify the application node] *******************************************
[1;30mtask path: /chaoslib/kubectl/cordon_drain_node.yaml:3[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod percona-6966c6db86-8lwfb -n app-percona-ns --no-headers -o custom-columns=:spec.nodeName", "delta": "0:00:01.213959", "end": "2018-11-26 11:56:21.364848", "rc": 0, "start": "2018-11-26 11:56:20.150889", "stderr": "", "stderr_lines": [], "stdout": "gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6", "stdout_lines": ["gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6"]}[0m

TASK [Record the application node name] ****************************************
[1;30mtask path: /chaoslib/kubectl/cordon_drain_node.yaml:11[0m
[0;32mok: [127.0.0.1] => {"ansible_facts": {"app_node": "gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6"}, "changed": false}[0m

TASK [Drain the application node] **********************************************
[1;30mtask path: /chaoslib/kubectl/cordon_drain_node.yaml:15[0m
[0;33mchanged: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl drain gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6 --ignore-daemonsets --force", "delta": "0:00:33.962611", "end": "2018-11-26 11:56:56.323069", "rc": 0, "start": "2018-11-26 11:56:22.360458", "stderr": "WARNING: Ignoring DaemonSet-managed pods: fluentd-gcp-v3.1.0-phftv, fluentd-forwarder-6xrvk, maya-io-agent-f7d8h, weave-scope-agent-24zc9, openebs-ndm-zr25m", "stderr_lines": ["WARNING: Ignoring DaemonSet-managed pods: fluentd-gcp-v3.1.0-phftv, fluentd-forwarder-6xrvk, maya-io-agent-f7d8h, weave-scope-agent-24zc9, openebs-ndm-zr25m"], "stdout": "node/gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6 cordoned\npod/litmus-percona-p9f2g-2z6j4 evicted\npod/openebs-provisioner-7c944cfd97-jmsh6 evicted\npod/openebs-snapshot-operator-77b785dbff-62dbs evicted\npod/percona-6966c6db86-8lwfb evicted\npod/cstor-sparse-pool-te9t-786dfc6d65-rvhnh evicted\npod/maya-apiserver-8c9c4df97-nxxnp evicted", "stdout_lines": ["node/gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6 cordoned", "pod/litmus-percona-p9f2g-2z6j4 evicted", "pod/openebs-provisioner-7c944cfd97-jmsh6 evicted", "pod/openebs-snapshot-operator-77b785dbff-62dbs evicted", "pod/percona-6966c6db86-8lwfb evicted", "pod/cstor-sparse-pool-te9t-786dfc6d65-rvhnh evicted", "pod/maya-apiserver-8c9c4df97-nxxnp evicted"]}[0m

TASK [Wait for application pod reschedule (evict)] *****************************
[1;30mtask path: /chaoslib/kubectl/cordon_drain_node.yaml:26[0m
[0;32mok: [127.0.0.1] => {"changed": false, "elapsed": 30, "path": null, "port": null, "search_regex": null, "state": "started"}[0m

TASK [Uncordon the application node] *******************************************
[1;30mtask path: /chaoslib/kubectl/cordon_drain_node.yaml:35[0m
[0;36mskipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}[0m

TASK [Verify AUT liveness post fault-injection] ********************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:96[0m
[0;33mchanged: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n app-percona-ns -l name=percona --no-headers -o custom-columns=:status.phase", "delta": "0:00:01.170523", "end": "2018-11-26 11:57:29.103065", "rc": 0, "start": "2018-11-26 11:57:27.932542", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}[0m

TASK [Get the liveness pods] ***************************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:110[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod -n litmus -l liveness=node-chaos -o=custom-columns=NAME:\".metadata.name\" --no-headers", "delta": "0:00:00.848462", "end": "2018-11-26 11:57:30.325431", "rc": 0, "start": "2018-11-26 11:57:29.476969", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}[0m

TASK [Checking status of liveness pods] ****************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:114[0m

TASK [set_fact] ****************************************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:124[0m
[0;32mok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}[0m

TASK [Uncordon the application node] *******************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:134[0m
[0;33mchanged: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl uncordon gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6", "delta": "0:00:00.970267", "end": "2018-11-26 11:57:31.987359", "rc": 0, "start": "2018-11-26 11:57:31.017092", "stderr": "", "stderr_lines": [], "stdout": "node/gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6 uncordoned", "stdout_lines": ["node/gke-e2e-cstor-dev-1-default-pool-224a4771-k4z6 uncordoned"]}[0m

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
[1;30mtask path: /percona/chaos/drain_node/test.yml:144[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "checksum": "1eb92f93de27bd2d562130a84d6c2a5d3672e533", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "689036f9df35339840e059b983edc89d", "mode": "0644", "owner": "root", "size": 434, "src": "/root/.ansible/tmp/ansible-tmp-1543233452.2-266551253728510/source", "state": "file", "uid": 0}[0m

TASK [Apply the litmus result CR] **********************************************
[1;30mtask path: /percona/chaos/drain_node/test.yml:155[0m
[0;33mchanged: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.088895", "end": "2018-11-26 11:57:35.212788", "failed_when_result": false, "rc": 0, "start": "2018-11-26 11:57:34.123893", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/node-failure-drain configured", "stdout_lines": ["litmusresult.litmus.io/node-failure-drain configured"]}[0m
[0;34mMETA: ran handlers[0m
[0;34mMETA: ran handlers[0m

PLAY RECAP *********************************************************************
[0;33m127.0.0.1[0m                  : [0;32mok=18  [0m [0;33mchanged=12  [0m unreachable=0    failed=0   

```

File: apps/percona/chaos/drain_node/run_litmus_test.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      volumes:
+        - name: kubeconfig
+          configMap:
+            name: kubeconfig
+        - name: logs
+          hostPath:
+            path: /mnt/chaos/node-failure-drain
+            type: ""
+
+      tolerations:
```
Comment (ksatchit) [inline apps/percona/chaos/drain_node/run_litmus_test.yml]: any particular reason in placing tolerations here @dargasudarshan ?

Comment (dargasudarshan) [reply apps/percona/chaos/drain_node/run_litmus_test.yml]: No particular reason @ksatchit Just followed k8s docs example.

**Comment by ksatchit** [comment]
Thanks @dargasudarshan - do you think it would be good to add the mysql persistence check here as well?

**Comment by dargasudarshan** [comment]
Yeah sure @ksatchit I will add mysql persistence check util and set job env by default to disable so that test can be generic to run on any application.

File: apps/percona/chaos/drain_node/run_litmus_test.yml (hunk 1 above)
Comment (dargasudarshan) [reply apps/percona/chaos/drain_node/run_litmus_test.yml]:
```
kubectl taint node gke-e2e-cstor-dev-1-default-pool-224a4771-gdmk infra-aid=observer:NoSchedule                 
node "gke-e2e-cstor-dev-1-default-pool-224a4771-gdmk" tainted
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona/liveness (strong-eon-153112)$ watch kubectl get pods -n app-percona-ns -o wide                                                                
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona/liveness (strong-eon-153112)$ cd ..
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona (strong-eon-153112)$ cd chaos/drain_node/
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona/chaos/drain_node (strong-eon-153112)$ ls
run_litmus_test.yml  test_vars.yml  test.yml
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona/chaos/drain_node (strong-eon-153112)$ vi run_litmus_test.yml
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona/chaos/drain_node (strong-eon-153112)$ kubectl create -f run_litmus_test.yml
job.batch "node-failure-drain-cl57b" created
sudarshan_darga@cloudshell:~/pr/litmus/apps/percona/chaos/drain_node (strong-eon-153112)$ kubectl get pods -n litmus -o wide
NAME                             READY     STATUS              RESTARTS   AGE       IP            NODE
litmus-percona-k8kkj-w6bn6       0/1       Completed           0          8m        10.20.22.14   gke-e2e-cstor-dev-1-default-pool-224a4771-jq6g
node-failure-drain-cl57b-6kbjr   0/2       ContainerCreating   0          10s       <none>        gke-e2e-cstor-dev-1-default-pool-224a4771-gdmk
```

**Comment by dargasudarshan** [comment]
Testing the changes.

**Comment by vibhor995** [comment]
[drain.log](https://github.com/openebs/litmus/files/2623881/drain.log)

Logs 
@dargasudarshan

**Pull request merged**